pandas/_libs/ops.pyx

```diff
@@ -0,0 +1,236 @@
+            x = values[i]
+            if checknull(x):
+                result[i] = False
+            elif isnull_val:
+                result[i] = False
+            else:
+                result[i] = PyObject_RichCompareBool(x, val, flag)
+
+    return result.view(bool)
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def vec_compare(ndarray[object] left, ndarray[object] right, object op):
+    cdef:
+        Py_ssize_t i, n = len(left)
+        ndarray[uint8_t, cast=True] result
+        int flag
+
+    if n != len(right):
+        raise ValueError('Arrays were different lengths: %d vs %d'
+                         % (n, len(right)))
+
+    if op is operator.lt:
+        flag = Py_LT
+    elif op is operator.le:
+        flag = Py_LE
+    elif op is operator.gt:
+        flag = Py_GT
+    elif op is operator.ge:
+        flag = Py_GE
+    elif op is operator.eq:
+        flag = Py_EQ
+    elif op is operator.ne:
+        flag = Py_NE
+    else:
+        raise ValueError('Unrecognized operator')
+
+    result = np.empty(n, dtype=bool).view(np.uint8)
+
+    if flag == Py_NE:
+        for i in range(n):
+            x = left[i]
+            y = right[i]
+
+            if checknull(x) or checknull(y):
+                result[i] = True
+            else:
+                result[i] = PyObject_RichCompareBool(x, y, flag)
+    else:
+        for i in range(n):
+            x = left[i]
+            y = right[i]
+
+            if checknull(x) or checknull(y):
+                result[i] = False
+            else:
+                result[i] = PyObject_RichCompareBool(x, y, flag)
+
+    return result.view(bool)
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def scalar_binop(ndarray[object] values, object val, object op):
+    cdef:
+        Py_ssize_t i, n = len(values)
+        ndarray[object] result
+        object x
+
+    result = np.empty(n, dtype=object)
+    if _checknull(val):
+        result.fill(val)
+        return result
+
+    for i in range(n):
+        x = values[i]
+        if _checknull(x):
+            result[i] = x
+        else:
+            result[i] = op(x, val)
+
+    return maybe_convert_bool(result)
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def vec_binop(ndarray[object] left, ndarray[object] right, object op):
+    cdef:
+        Py_ssize_t i, n = len(left)
+        ndarray[object] result
+
+    if n != len(right):
+        raise ValueError('Arrays were different lengths: %d vs %d'
+                         % (n, len(right)))
+
+    result = np.empty(n, dtype=object)
+
+    for i in range(n):
+        x = left[i]
+        y = right[i]
+        try:
+            result[i] = op(x, y)
+        except TypeError:
+            if _checknull(x):
+                result[i] = x
+            elif _checknull(y):
+                result[i] = y
+            else:
+                raise
+
+    return maybe_convert_bool(result)
+
+
+def maybe_convert_bool(ndarray[object] arr,
```

Comment:
It's used twice in this module and once in `io.parsers`.  Maybe if/when there is a _libs.cast it might make sense to put it there, but until then this seems like the right home for it.